packages/cozy-client/src/CozyClient.js

```diff
@@ -198,42 +198,54 @@ class CozyClient {
       event: TRIGGER_CREATION
     }
   ) {
-    if (!document._type) return document
-
-    const schema = this.schema.getDoctypeSchema(document._type)
-    if (!schema.cozyMetadata) return document
-
-    const enriched = {
-      ...document
-    }
-    const metadata = Object.entries(schema.cozyMetadata)
-
-    for (const [
-      metadataName,
-      { trigger, value, useCurrentDate = false }
-    ] of metadata) {
-      const shouldSetMetadata =
-        options.event === TRIGGER_CREATION ||
-        (options.event === TRIGGER_UPDATE && trigger === TRIGGER_UPDATE)
-      if (!shouldSetMetadata) continue
-
-      let newValue
-      const metadataPath = `cozyMetadata.${metadataName}`
-
-      const shouldAccumulateValues = Array.isArray(value)
-      if (shouldAccumulateValues) {
-        const prevValue = get(document, metadataPath, [])
-        newValue = uniqBy([...prevValue, ...value])
-      } else if (useCurrentDate) {
-        newValue = new Date().toISOString()
-      } else {
-        newValue = get(document, metadataPath, value)
+    if (this.metadata === undefined) return document
+
+    const {
+      doctypeVersion,
+      slug,
+      sourceAccount = null,
+      version
+    } = this.metadata
+
+    const now = new Date().toISOString()
+
+    let cozyMetadata = get(document, 'cozyMetadata', {})
+    if (options.event === TRIGGER_CREATION) {
+      cozyMetadata = {
+        ...cozyMetadata, // custom metadata that are set by the app
+        doctypeVersion,
+        createdByApp: slug,
+        sourceAccount,
+        createdAt: now,
+        createdByAppVersion: version,
+        updatedAt: now,
+        updatedByApps: [
+          {
+            date: now,
+            slug
+          }
+        ]
+      }
+    } else if (options.event === TRIGGER_UPDATE) {
+      cozyMetadata = {
+        // add doctypeVersion and sourceAccount if they don't exist
+        doctypeVersion,
+        sourceAccount,
+        ...cozyMetadata,
+        updatedAt: now,
+        updatedByApps: [
+          ...get(document, 'cozyMetadata.updatedByApps', []).filter(
+            info => info.slug !== slug
+          ),
+          { date: now, slug }
```

Comment:
> By convention, the first app (slug) listed in updatedByApps is the one which did the last update.

… so we should move the `slug` in front of the spread.

Reply:
Once we add the date in each object, this convention is to be removed. Let's simplify things

Reply:
I removed this sentence here: https://github.com/cozy/cozy-doctypes/pull/79/files
But it may be useful (but maybe dangerous) to always have the last update first since we use an array